### PR TITLE
mgr/progress: optimize global recovery module

### PR DESCRIPTION
**Pull request by kamoltat**
Instead of traversing through each PGs for active+clean state. 
We traverse through the PGs in
ActivePyModules.cc which is in C++ and use get
method to directly get active+clean PG nums, as
well as number of PGs.

Signed-off-by: Kamoltat <ksirivad@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
jenkins test docs

File: src/mgr/ActivePyModules.cc [hunk 1]
```diff
@@ -432,6 +432,28 @@ PyObject *ActivePyModules::get_python(const std::string &what)
       mgr_map.dump(&f);
     });
     return f.get();
+  } else if (what == "active_clean"){
```
Comment (jdurgin) [inline src/mgr/ActivePyModules.cc:435]: spacing looks odd on github (not sure if it's right) but the logic looks correct

**Comment by kamoltat** [comment]
jenkins test docs

**Comment by kamoltat** [comment]
jenkins retest this please

**Comment by kamoltat** [comment]
jenkins retest this please

**Comment by kamoltat** [comment]
jenkins retest this please

**Comment by kamoltat** [comment]
https://pulpito.ceph.com/ksirivad-2020-11-11_06:04:09-rados:mgr-wip-mgr-progress-global-efficiency-distro-basic-smithi/

**Comment by kamoltat** [comment]
jenkins retest this please

**Comment by kamoltat** [comment]
https://pulpito.ceph.com/ksirivad-2020-11-16_04:02:02-rados-wip-mgr-progress-global-efficiency-distro-basic-smithi/

**Comment by kamoltat** [comment]
reran: dead/fail jobs https://pulpito.ceph.com/ksirivad-2020-11-20_07:21:48-rados-wip-mgr-progress-global-efficiency-distro-basic-smithi/

Seems like the tests went well

**Review by jdurgin — APPROVED** [review]

File: src/mgr/ActivePyModules.cc [hunk 2]
```diff
@@ -441,6 +441,28 @@ PyObject *ActivePyModules::get_python(const std::string &what)
...
+      uint64_t num_active_clean_pg = 0;
+      for (auto p = pg_map.num_pg_by_state.begin();
+	  p != pg_map.num_pg_by_state.end();
+	  ++p) {
+	  if (p->first & PG_STATE_ACTIVE) {
+	     if (p->first & PG_STATE_CLEAN){
+	     	num_active_clean_pg += p->second;
+		}
+	  }
+      }
```
Comment (tchaikov) [inline src/mgr/ActivePyModules.cc:462]:
```suggestion
  for (auto [state, num] : pg_map.num_pg_by_state) {
    if (state & (PG_STATE_ACTIVE | PG_STATE_CLEAN) {
      num_active_clean_pg += num;
    }
  }
```

Comment (kamoltat) [reply src/mgr/ActivePyModules.cc]: thanks! that's much more eloquent

**Comment by tchaikov** [comment]
@kamoltat needs rebase.

**Comment by tchaikov** [comment]
@kamoltat i have following build failure 
```
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-572-gccc67fce/rpm/el8/BUILD/ceph-17.0.0-572-gccc67fce/src/mgr/ActivePyModules.cc: In member function 'PyObject* ActivePyModules::get_python(const string&)':
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-572-gccc67fce/rpm/el8/BUILD/ceph-17.0.0-572-gccc67fce/src/mgr/ActivePyModules.cc:485:15: error: 'tstate' was not declared in this scope
         [&f, &tstate](const PGMap &pg_map) {
               ^~~~~~
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/17.0.0-572-gccc67fce/rpm/el8/BUILD/ceph-17.0.0-572-gccc67fce/src/mgr/ActivePyModules.cc:485:15: note: suggested alternative: '_xstate'
         [&f, &tstate](const PGMap &pg_map) {
               ^~~~~~
               _xstate
```
while compiling my batch when testing this PR. could you fix the build failure and repush?

**Comment by kamoltat** [comment]
jenkins test docs

**Comment by jdurgin** [comment]
@kamoltat this was looking good, seems like it just needs a rebase

**Pull request merged**